module: add API for interacting with source maps (#31132)
nodejs/node: doc/api/source_map.md

# Source Map

<!--introduced_in=REPLACEME-->

> Stability: 1 - Experimental

<!--name=source_map-->

> Exposes an API for interacting with Node.js' internal cache of source maps.

Reviewer: If this is only useful for interacting with the internal cache, why do we need to expose the SourceMap constructor?

The source maps at the moment are essentially mapping between source positions, which are essentially tuples of (filename, line, column) or (filename, line, column, name) (target-only), why do we require users to pass an error to the API to get a source map first before fetching the mapping, instead of just using the a straight-forward API with a signature like `(filename, line, column) => {filename, line, column, name?}` ? IIUC aren't the error-rekeying only useful for `Module._load`, which is written in JS and can't be source-mapped anyway?

Author: > If this is only useful for interacting with the internal cache, why do we need to expose the SourceMap constructor?

There are other use cases, and I thought it might be worthwhile to give authors access to the class:

1. it means for their own unit testing, they can instantiate a `SourceMap` instance with a Source Map V3 payload; since we don't allow folks to directly access the cache, I think this would be beneficial.
2. there are a few other methods I imagine we might expose, `findEntryReversed`, which was left out for now; methods for interacting with `1:many` source maps (which potentially get more complex, and require more helpers); I think it might be an ugly abstraction putting these all on the cache.
3. keep in mind the SourceMap instance is almost directly from V8, and compliant with the Source Map V3 specification; putting it in the hands of tooling authors will make it more likely that this feature sees adoption, I imagine folks might find use-cases I haven't thought of in Node.js' naive initial implementation of SourceMaps.

> IIUC aren't the error-rekeying only useful for Module._load, which is written in JS and can't be source-mapped anyway?

Because of the WeakMap we ended up using to cache source maps against modules, keying on `error` is necessary if an exception happened while loading a module, so cases like this:

```js
  Error.prepareStackTrace = (error, trace) => {
    const throwingRequireCallSite = trace[0];
    if (throwingRequireCallSite.getFileName().endsWith('typescript-throw.js')) {
      sourceMap = findSourceMap(throwingRequireCallSite.getFileName(), error);
      callSite = throwingRequireCallSite;
    }
  };
  try {
    // Require a file that throws an exception, and has a source map.
    require('../fixtures/source-map/typescript-throw.js');
  } catch (err) {
    err.stack; // Force prepareStackTrace() to be called.
  }
```

☝️ Given that one of the common use cases of source maps is prettifying stack traces, I think this will actually come up pretty often.

Author: With regards to:

> IIUC aren't the error-rekeying only useful for Module._load

I agree that it's not an intuitive API surface to ask the user to provide an error object, I wonder if it would be worth complicating the data-structure we store the source maps in a bit more (perhaps a secondary index?) to remove this burden from the user.

Reviewer: > it means for their own unit testing, they can instantiate a SourceMap instance with a Source Map V3 payload; since we don't allow folks to directly access the cache, I think this would be beneficial.

How would they use this in tests? I don't think there is an API that takes a source map object and alters the internal cache? To me it's hard to imagine how one would write a test with this other than testing Node's own implementation of the source map parsing. If one wants to make sure that they emits the correct source map, I assume they would just write files to a temporary directory and check if they can get the correct positions from the Node after loading these files.

> keep in mind the SourceMap instance is almost directly from V8

I am confused - what exactly comes from V8? The source maps are emitted by the users, we get the URI by parsing the file content with regex ourselves, and load them by reading from the file system / decoding the data URI ourselves? The SourceMap class is just a wrapper on top of the data we parse and read ourselves?

> I think this will actually come up pretty often.

So `error` is only necessary when the `source_map.findSourceMap` may be invoked when handling an error that occurs during module loading? It still seems somewhat awkward to me that we require the user to do this due to the limitation of the current implementation, at least we should explain in the docs about the reason (and what would happen if they forget to pass it)

Author: > I am confused - what exactly comes from V8?

The class `SourceMap` is a faithful implementation of the [Source Map Revision 3 Proposal](https://sourcemaps.info/spec.html), [taken from the V8 codebase](https://github.com/v8/v8/blob/4b9b23521e6fd42373ebbcb20ebe03bf445494f9/tools/SourceMap.js).

> The source maps are emitted by the users, we get the URI by parsing the file content with regex ourselves, and load them by reading from the file system / decoding the data URI ourselves? 

Many tooling authors have the use case of consuming source maps coming from a variety of locations, Node.js' cache I would expect to be just one...

As an example, take [puppeteer](https://github.com/puppeteer/puppeteer) which runs in the browser, _but is orchestrated by Node.js_; I would like the user to be able to take the SourceMaps fetched from code executed via the puppeteer harness, instantiate a `SourceMap` instance, and interact with it identically to as if they'd fetched it from Node.js' cache.

> How would they use this in tests? I don't think there is an API that takes a source map object and alters the internal cache? To me it's hard to imagine how one would write a test with this other than testing Node's own implementation of the source map parsing

I was picturing writing a tool like Babel, that generates source-maps; if they want to write unit tests to ensure these payloads work properly with Node.js' implementation, because the payload format is standardized, they can simply instantiate a `SourceMap` class (there's no need for it to have ever been in Node.js' cache) -- this makes it easier to write targeted unit tests, without going through the dance of requiring a module in Node.js, and forcing it into cache.

> So error is only necessary when the source_map.findSourceMap may be invoked when handling an error that occurs during module loading? 

Correct, the `error` is required for exceptions that occur during the loading process; I'm happy to document this further.

I also think it's worth adding a few TODO comments, because we can get rid of this requirement as soon as [WeakReferences are available in Node.js](https://github.com/tc39/proposal-weakrefs); I think the source-map cache can be simplified significantly with an interable WeakMap.

Reviewer: Thanks for the explanation.

> taken from the V8 codebase.

Shouldn't the source of truth be https://cs.chromium.org/chromium/src/third_party/devtools-frontend/src/front_end/sdk/SourceMap.js ? It seems v8's copy has been outdated (also, it is only used in the tick processor, though TBH I have never use the tick processor with source maps before..)

Author: @joyeecheung seems like I should read through the changes that have landed in `front_end/sdk/SourceMap.js` and bring any important ones over.

> Exposes an API for interacting with Node.js' internal cache of source maps.

Reviewer 2: Utter micro-nit but it might be easiest for readers if we don't use the possessive in this instance:

```suggestion
Exposes an API for interacting with the internal cache of source maps.
```

...or...

```suggestion
Exposes an API for interacting with the Node.js internal cache of source maps.
```

Or even:

```suggestion
Exposes an API for interacting with the cache of source maps.
```

To enable source map parsing, Node.js must be run with the flag
[`--enable-source-maps`][], or with code coverage enabled by setting
[`NODE_V8_COVERAGE=dir`][].

```js
const source_map = require('source_map');
```

> The APIs and implementation are subject to change at any time.

Reviewer 3: I don't think this sentence is necessary given the `Experimental` status notification above.
```suggestion
```

## `source_map.findSourceMap(path[, error])`
<!-- YAML
added: REPLACEME
-->

* `path` {string}
* `error` {Error}
* Returns: {source_map.SourceMap}

> `path` should be the resolved path for the file that you wish to fetch a

Reviewer 3: Style nit: Please avoid using `you` in the docs
```suggestion
`path` is the resolved path for the file for which a corresponding source map should be fetched.
```

> corresponding source map for.

Reviewer 3: ```suggestion
```

`error` if you are interacting with source maps in an exceptional flow, e.g.,
having overridden [`Error.prepareStackTrace(error, trace)`][], you should pass

> the error instance as a second parameter to `findSourceMap`.

Reviewer 3: What's the expected behavior here when `--enable-source-maps` is *not* enabled.

Author: If `--enable-source-maps` or `NODE_V8_COVERAGE` have not been set, then any call to `findSourceMap` will fail to find a source map corresponding to the path, and `null` will be the return value.

> the error instance as a second parameter to `findSourceMap`.

Reviewer 3: Please reword to avoid use of `you`
